fix: gpu-handling for made and leakage correction

janfb opened the pull request:
during GPU tests a couple of bugs appeared, mostly for the new `CategoricalMADE`. 

I overloaded the `to` methods for the mixed density networks to avoid moving the `masks` to the device explicitly in every `forward` call.

manuelgloeckler commented on sbi/inference/posteriors/direct_posterior.py:
```diff
@@ -384,7 +384,7 @@ def acceptance_at(x: Tensor) -> Tensor:
                     )
                 },
-            )[1]
+            )[1].to(self._device)
```
Why is this necessary? The proposal is `posterior_estimator,` which should be on the device and thus produce samples from it. I guess this is the "acceptance_rate", but this should be computed from the samples and thus also be on device.

manuelgloeckler commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -159,6 +159,11 @@ def sample(
 
+    def to(self, device):
+        """Override `to()` to move self.mask along with the model."""
+        self.mask = self.mask.to(device)
```
Usually, one registers non-differentiable/static but necessary parts as "Buffers" in PyTorch modules. Buffers will be automatically moved onto the device i.e. in the init do `self.register_buffer("mask", mask)` instead.

janfb replied:
oh yes, very good point! will change it to that.

manuelgloeckler commented on sbi/neural_nets/estimators/categorical_net.py:
```diff
@@ -231,3 +236,8 @@ def loss(self, input: Tensor, condition: Tensor, **kwargs) -> Tensor:
 
         return -self.log_prob(input.unsqueeze(0), condition)[0]
+
+    def to(self, device):
```
This should not be necessary. Might be a problem within the `__init__`.
```python
super().__init__(
            net=net, input_shape=input_shape, condition_shape=condition_shape
        )
self.net = net           # Why again override net?
self.num_categories = net.num_categories
```

manuelgloeckler commented on sbi/neural_nets/estimators/mixed_density_estimator.py:
```diff
@@ -198,6 +198,13 @@ def loss(self, input: Tensor, condition: Tensor, **kwargs) -> Tensor:
         """
         return -self.log_prob(input.unsqueeze(0), condition)[0]
 
+    def to(self, device):
```
Same as the above. This should not be necessary.

manuelgloeckler reviewed (commented):
Thanks for fixing, while it does solve the problem I added a few comments to the changes on the estimators. I think if one properly registers the other tensors used in the forward pass, the errors should be also fixed.

janfb commented:
> Thanks for fixing, while it does solve the problem I added a few comments to the changes on the estimators. I think if one properly registers the other tensors used in the forward pass, the errors should be also fixed.

Thanks! your suggestions make sense and I pushed it. I cannot test it with a GPU at the moment though.

manuelgloeckler commented:
I can test it on my laptop, I guess all GPU tests should be sufficient.

janfb commented:
> I can test it on my laptop, I guess all GPU tests should be sufficient.

it's just: 
- `pytest -m "gpu" tests/mnle_test.py` 
- `pytest -m "gpu" tests/analysis_test.py::test_analysis_modules`

manuelgloeckler commented:
This test does fail:
- pytest -m gpu tests/inference_on_device_test.py::test_training_and_mcmc_on_device[gaussian-gpu-gpu-NRE_C-resnet-nuts_pymc]

Due to
```
device = 'cuda:0', prior = MultivariateNormal(loc: torch.Size([2]), covariance_matrix: torch.Size([2, 2]))

    def check_if_prior_on_device(
        device: Union[str, torch.device], prior: Optional[Any] = None
    ) -> None:
        """Try to sample from the prior, and check that the returned data is on the correct
        trainin device. If the prior is `None`, simplys pass.
    
        Args:
            device: target torch training device
            prior: any simulator outputing torch `Tensor`
        """
        if prior is None:
            pass
        else:
            prior_device = prior.sample((1,)).device
            training_device = torch.zeros(1, device=device).device
>           assert prior_device == training_device, (
                f"Prior device '{prior_device}' must match training device "
                f"'{training_device}'. When training on GPU make sure to "
                "pass a prior initialized on the GPU as well, e.g., "
                "prior = torch.distributions.Normal"
                "(torch.zeros(2, device='cuda'), scale=1.0)`."
            )
E           AssertionError: Prior device 'cpu' must match training device 'cuda:0'. When training on GPU make sure to pass a prior initialized on the GPU as well, e.g., prior = torch.distributions.Normal(torch.zeros(2, device='cuda'), scale=1.0)`.
```

Not sure why.

manuelgloeckler commented:
Ahh, now it xfails as intended. I was on an outdated pymc version, will run all tests on a fresh installation and approve this PR once they pass.

manuelgloeckler reviewed (approved):
All GPU tests pass now.
